### PR TITLE
Initial version of a statistics page

### PR DESCRIPTION
**Pull request by feeds**
First version of #271

### CODE REVIEW

**Comment by Baschdl** [comment]
We need a better user permission system for this. @maltezacharias is on it

**Comment by feeds** [comment]
Well, we should open another issue for this, right?

I can easily add another group of 'statsviewers', what is @maltezacharias onto? 😅

**Comment by Baschdl** [comment]
Yes, he opened #466. Shouldn't prevent merging this PR

File: backend/apps/accounts/db_stats.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        return (
+            _("Registrierte Helfende"),
+            Student.objects.filter(user__validated_email=True).count(),
+        )
+
+    @stat_count.register
+    def deactivated_accounts(self):
+        return (
+            _("Anzahl deaktivierter Helfenden"),
+            Student.objects.filter(is_activated=True).count(),
```
Comment (Baschdl) [inline backend/apps/accounts/db_stats.py]: Shouldn't that be `False`?
```suggestion
            Student.objects.filter(is_activated=False).count(),
```

File: backend/apps/accounts/db_stats.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    # - helfende pro bundesland und großstadt
+
+    ## Contact stats
+    @stat_count.register
+    def students_contacted_by_hospital(self):
+        return (_("Kontaktanfragen an Helfende"), EmailToSend.objects.filter(was_sent=True).count())
+
+    @stat_count.register
+    def hospitals_contacted_by_students(self):
+        return (_("Kontaktanfragen an Institutionen"), EmailToHospital.objects.filter().count())
```
Comment (Baschdl) [inline backend/apps/accounts/db_stats.py]:
```suggestion
        return (_("Kontaktanfragen an Institutionen"), EmailToHospital.objects.count())
```

File: backend/apps/accounts/db_stats.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+            _("Anzahl deaktivierter Helfenden"),
+            Student.objects.filter(is_activated=True).count(),
+        )
+
+    # todo:
+    # - helfende pro bundesland und großstadt
+
+    ## Contact stats
+    @stat_count.register
+    def students_contacted_by_hospital(self):
```
Comment (Baschdl) [inline backend/apps/accounts/db_stats.py]:
```suggestion
    def emails_to_students(self):
```
The method sounds like the number of unique students which where contacted

File: backend/apps/accounts/db_stats.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        return (_("Kontaktanfragen an Helfende"), EmailToSend.objects.filter(was_sent=True).count())
+
+    @stat_count.register
+    def hospitals_contacted_by_students(self):
```
Comment (Baschdl) [inline backend/apps/accounts/db_stats.py]:
```suggestion
    def emails_to_hospitals(self):
```

**Review by Baschdl — CHANGES_REQUESTED** [review]
Did you leave out "Deactivated Institutions" on purpose?

**Comment by feeds** [comment]
I did not know they could deactivate themselves. I can add it if required.

**Comment by Baschdl** [comment]
> I did not know they could deactivate themselves. I can add it if required.

Now I'm not sure either. Isn’t that part of the basic user functionality? If it's not, why can’t they deactivate themselves if they found someone for now but need more help later?

Can’t test it locally right now.

File: backend/apps/accounts/db_stats.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+            except StopIteration:
+                count = 0
+            res.append((form_labels[t], count))
+        return res
+
+    def all_stats(self):
+        results = [m(self) for m in self.stat_count]
+        for m in self.stat_list:
+            for r in m(self):
+                results.append(r)
```
Comment (Baschdl) [inline backend/apps/accounts/db_stats.py]:
```suggestion
            results.extend(m(self))
```

**Review by Baschdl — APPROVED** [review]
Otherwise it's fine

**Comment by feeds** [comment]
thank youuu :)

File: backend/apps/accounts/db_stats.py [hunk 6]
```diff
@@ -19,7 +19,7 @@ class DataBaseStats:
     stat_count = RegisterList()
     stat_list = RegisterList()
 
-    # TODO: last X days?
+    # TODO: last X days? # noqa: T003
```
Comment (bjrne) [inline backend/apps/accounts/db_stats.py:22]: This is not the idea @Baschdl :laughing: 
Either its an idea that isn't worth pursuing, or it is a TODO and should get its own issue for keeping track of it :D

File: backend/apps/accounts/db_stats.py [hunk 7]
```diff
@@ -46,7 +46,8 @@ def deactivated_accounts(self):
             Student.objects.filter(is_activated=False).count(),
         )
 
-    # TODO: helfende pro bundesland und großstadt
+    # TODO: helfende pro bundesland und großstadt. Requires
+    # https://github.com/match4everyone/match4healthcare/issues/492
```
Comment (bjrne) [inline backend/apps/accounts/db_stats.py:50]: Same here, the issue is supposed to be about the TODO itself. In this way, nobody will ever check this again. If they were linked on github, one could see more easily whether the dependency issue is closed.

**Review by bjrne — COMMENTED** [review]
Nooooooooo :D

File: backend/apps/accounts/db_stats.py (hunk 6 above)
Comment (Baschdl) [reply backend/apps/accounts/db_stats.py:22]: That's a PO issue where we just have to wait for input from the rest. Not sure if opening an issue for all TODOs is a good idea. I fear that the issues will get completely messy

File: backend/apps/accounts/db_stats.py (hunk 7 above)
Comment (Baschdl) [reply backend/apps/accounts/db_stats.py:50]: We can reference this as `dependent` in 492 but I wouldn't open a separate issue

**Pull request merged**